File: python/atelier_bench/rules/schedule_gap.py
"""schedule-gap rule (nominal: subclasses RuleContract)."""

from __future__ import annotations

from typing import TYPE_CHECKING

from ..contracts.rule_contract import RuleContract

if TYPE_CHECKING:
    from ..dataset import Dataset


class ScheduleGapRule(RuleContract):
    KEY = "schedule-gap"

    def __init__(self) -> None:
        self.key = self.KEY

    def evaluate(self, data: "Dataset") -> bool:
        return all(technician.next_slot() is not None for technician in data.technicians)

File: go/atelier/rules/schedule_gap.go
// Package rules carries the 48 rules: 24 NOMINAL rule structs in this package
// (each with a compile-time assertion against RuleContract and a value
// receiver) plus the 24 STRUCTURAL ad-hoc rule values from rules/adhoc, which
// assert nothing and satisfy the same method set by shape alone.
package rules

import (
	"atelier.example/lane/atelier/contracts"
	"atelier.example/lane/atelier/dataset"
)

// Compile-time proof against the NOMINAL contract, asserted on the VALUE
// type: ScheduleGapRule's methods take value receivers, so both ScheduleGapRule and
// *ScheduleGapRule satisfy RuleContract.
var _ contracts.RuleContract = ScheduleGapRule{}

// ScheduleGapRule is the schedule-gap rule.
type ScheduleGapRule struct{}

// Key is the rule identifier.
func (ScheduleGapRule) Key() string {
	return "schedule-gap"
}

// Evaluate reports whether the rule holds on this dataset.
func (r ScheduleGapRule) Evaluate(data *dataset.Dataset) bool {
	for _, technician := range data.Technicians {
		if technician.NextSlot() < 0 {
			return false
		}
	}
	return true
}

// init registers this rule.
func init() {
	Register(ScheduleGapRule{})
}
